// egg-ts-helper/dist/watcher.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const path_1 = tslib_1.__importDefault(require("path"));
const chokidar_1 = tslib_1.__importDefault(require("chokidar"));
const assert_1 = tslib_1.__importDefault(require("assert"));
const events_1 = require("events");
const utils = tslib_1.__importStar(require("./utils"));
const debug_1 = tslib_1.__importDefault(require("debug"));
const debug = debug_1.default('egg-ts-helper#watcher');
let generators;
class Watcher extends events_1.EventEmitter {
    constructor(options, helper) {
        super();
        this.helper = helper;
        this.throttleTick = null;
        this.throttleStack = [];
        this.init(options);
    }
    init(options) {
        const generatorName = options.generator || 'class';
        this.config = this.helper.config;
        this.name = options.name;
        this.generator = this.getGenerator(generatorName);
        this.options = Object.assign({ trigger: ['add', 'unlink'], generator: generatorName, pattern: '**/*.(ts|js)', watch: true }, this.generator.defaultConfig, options);
        const baseDir = options.path.replace(/\/|\\/, path_1.default.sep);
        this.dir = utils.getAbsoluteUrlByCwd(baseDir, this.config.cwd);
        this.dtsDir = path_1.default.resolve(this.config.typings, path_1.default.relative(this.config.cwd, this.dir));
    }
    destroy() {
        if (this.fsWatcher) {
            this.fsWatcher.close();
        }
        clearTimeout(this.throttleTick);
        this.throttleTick = null;
        this.throttleStack.length = 0;
        this.removeAllListeners();
    }
    // watch file change
    watch() {
        if (!this.options.watch) {
            return;
        }
        if (this.fsWatcher) {
            this.fsWatcher.close();
        }
        // glob only works with / in windows
        const watchGlob = path_1.default
            .join(this.dir, this.options.pattern)
            .replace(/\/|\\/g, '/');
        const watcher = chokidar_1.default.watch(watchGlob, this.config.watchOptions);
        // listen watcher event
        this.options.trigger.forEach(evt => {
            watcher.on(evt, this.onChange.bind(this));
        });
        // auto remove js while ts was deleted
        if (this.config.autoRemoveJs) {
            watcher.on('unlink', utils.removeSameNameJs);
        }
        this.fsWatcher = watcher;
    }
    // execute generator
    execute(file) {
        debug('execution %s', file);
        const options = this.options;
        let _fileList;
        const newConfig = Object.assign({}, this.options, { file, dir: this.dir, dtsDir: this.dtsDir, get fileList() {
                return _fileList || (_fileList = utils.loadFiles(this.dir, options.pattern));
            } });
        const result = this.generator(newConfig, this.config, this.helper);
        if (result) {
            this.emit('update', result, file);
        }
        return result;
    }
    // on file change
    onChange(filePath) {
        debug('file changed %s %o', filePath, this.throttleStack);
        if (!this.throttleStack.includes(filePath)) {
            this.throttleStack.push(filePath);
        }
        if (this.throttleTick) {
            return;
        }
        this.throttleTick = setTimeout(() => {
            while (this.throttleStack.length) {
                this.execute(this.throttleStack.pop());
            }
            this.throttleTick = null;
        }, this.config.throttle);
    }
    // get generator
    getGenerator(name) {
        const type = typeof name;
        const typeIsString = type === 'string';
        generators = generators || utils.loadModules(path_1.default.resolve(__dirname, './generators'));
        let generator = typeIsString ? generators[name] : name;
        if (!generator && typeIsString) {
            try {
                // try to load generator as module path
                const generatorPath = name.startsWith('.')
                    ? path_1.default.join(this.config.cwd, name)
                    : name;
                generator = require(generatorPath);
            }
            catch (e) {
                // do nothing
            }
        }
        // check esm default
        if (typeof generator.default === 'function') {
            generator.default.defaultConfig = generator.defaultConfig;
            generator = generator.default;
        }
        assert_1.default(typeof generator === 'function', `generator: ${name} not exist!!`);
        return generator;
    }
}
exports.default = Watcher;
